### PR TITLE
fix(metrics): Check more thoroughly for cardinality limiter defaults

### PR DESCRIPTION
**Pull request by ayirr7**
We ran into a scenario where the cardinality limit option for a specific use case had been created, but hadn't actually been set. This should help us to fall back to defaults in this scenario.

### CODE REVIEW

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py [hunk 1]
```diff
@@ -49,7 +49,15 @@ def _construct_quotas(use_case_id: UseCaseID) -> Optional[Quota]:
     """
 
     if use_case_id in USE_CASE_ID_CARDINALITY_LIMIT_QUOTA_OPTIONS:
-        quota_args = options.get(USE_CASE_ID_CARDINALITY_LIMIT_QUOTA_OPTIONS[use_case_id])
+        custom_card_limit = options.get(USE_CASE_ID_CARDINALITY_LIMIT_QUOTA_OPTIONS[use_case_id])
+
+        # check that the option is set to a properly-formatted cardinality limit
+        if len(custom_card_limit) == 3:
```
Comment (john-z-yang) [inline src/sentry/sentry_metrics/indexer/limiters/cardinality.py:55]: I think this should be
```python3
if not custom_card_limit:
    ...
```

Comment (john-z-yang) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py:55]: Actually never mind. I thought options is a map

Comment (ayirr7) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py:55]: I think the default value for the option is actually `{}` so the length would be a valid check ?

Comment (john-z-yang) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py:55]: How about just
```python
if custom_card_limit:
    ...
```
I feel like checking if length is 3 might not be very readable

Comment (ayirr7) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py]: Yeah, you're right. I think that would work but I didn't see it in time (I just changed it to check for len() > 0 which I think checks for the same thing). Let me know if you want me to change it to your suggestion?

**Comment by john-z-yang** [comment]
This looks good to me. However, I really do think we need to add some sort of shout when falling back to
```python
options.get(
    "sentry-metrics.cardinality-limiter.limits.generic-metrics.per-org"
)
```
Because whoever onboarded the new use case definitely did something wrong

**Comment by ayirr7** [comment]
@john-z-yang Another point I wanted your thoughts on. Should we consider this scenario (a config exists, but isn't set) a bad state and emit some sort of warning? Thoughts on what kind of warning?

**Comment by ayirr7** [comment]
> I really think we need to shout when falling back to

Agreed. Will make this change for both the cardinality rate and writes rate limiters.

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py [hunk 2]
```diff
@@ -49,7 +49,15 @@ def _construct_quotas(use_case_id: UseCaseID) -> Optional[Quota]:
...
     if use_case_id in USE_CASE_ID_CARDINALITY_LIMIT_QUOTA_OPTIONS:
-        quota_args = options.get(USE_CASE_ID_CARDINALITY_LIMIT_QUOTA_OPTIONS[use_case_id])
+        custom_card_limit = options.get(USE_CASE_ID_CARDINALITY_LIMIT_QUOTA_OPTIONS[use_case_id])
+
+        # check that the option is set to something
+        if len(custom_card_limit) > 0:
+            quota_args = custom_card_limit
+        else:
+            quota_args = options.get(
+                "sentry-metrics.cardinality-limiter.limits.generic-metrics.per-org"
```
Comment (nikhars) [inline src/sentry/sentry_metrics/indexer/limiters/cardinality.py:59]: Nit: Lets define a constant for the default option to make code readability easier.

**Review by nikhars — COMMENTED** [review]
Please add a test case for this scenario.

**Comment by ayirr7** [comment]
@john-z-yang and I talked about it a bit, and we're thinking another possible solution would be to just make sure that the default values for these options are ones that are valid for the indexer (i.e. not an empty list `[]` which is the current default). Instead, we could set some dummy quota values into that list that (for example) permit all metrics from that particular use case.

**Comment by lynnagara** [comment]
Is it worth also handling case where the user provides any invalid formatted configuration? Unfortunately it is easy to do, and we may not want to crash the indexer in that scenario either.

**Pull request closed without merging**